prompt_toolkit/interface.py

```diff
@@ -1159,3 +1166,73 @@ def add_reader(self, fd, callback):
 
     def remove_reader(self, fd):
         self.cli.eventloop.remove_reader(fd)
+
+
+class _suppress_signal_exceptions(object):
+    """
+    Context manager that replaces all signal handlers by a wrapper that
+    suppresses exceptions from the original signal handlers.
+
+    Instead, when an exception was raised by the original signal handler, this
+    is stored and the `notify` callback is triggered. When later
+    `raise_exceptions` is called, the captured exception is raised.
+    """
+    def __init__(self, notify):
+        assert callable(notify)
+        self.notify = notify
+
+        # Map signal number to original handler.
+        self._original_handlers = {}
+        self.captured_exceptions = []
+
+    def __enter__(self):
+        # Only replace signals in the main thread.
+        # (It doesn't make sense to do this from other threads, as they don't
+        # receive signals anyway.)
+        if isinstance(threading.current_thread(), threading._MainThread):
+            # Replace each signal handler.
+            for n in self._get_signal_numbers():
+                try:
+                    original_handler = signal.getsignal(n)
+                except ValueError:  # Signal number out of range.
+                    pass
+                else:
+                    if original_handler:
```

Comment:
This will be True for ignored signals: `SIG_IGN` is like 1, but is not callable. We probably don't want to modify ignored signals, and possibly not signals with the 'default action' (which means killing the program for most signals).